File: apps/api/src/main/kotlin/com/dasida/api/notification/Notifications.kt
package com.dasida.api.notification

import com.fasterxml.jackson.annotation.JsonIgnore
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.Index
import jakarta.persistence.Table
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.http.HttpStatus
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.server.ResponseStatusException
import com.dasida.api.security.AuthUser
import java.time.Instant
import java.util.UUID

/** 알림 타입. 이번 PR 범위. */
object NotificationType {
    const val POST_COMMENT_CREATED = "POST_COMMENT_CREATED"
    const val CAMPAIGN_COMMENT_CREATED = "CAMPAIGN_COMMENT_CREATED"
    const val CAMPAIGN_JOINED = "CAMPAIGN_JOINED"
    const val CAMPAIGN_PARTICIPATION_REMOVED = "CAMPAIGN_PARTICIPATION_REMOVED"
}

/**
 * 사용자별 알림. userId 는 수신자이며 응답에 노출하지 않는다(@JsonIgnore).
 * 정렬은 seq DESC, id ASC. readAt == null 이면 unread.
 */
@Entity
@Table(
    name = "notifications",
    indexes = [
        Index(name = "idx_notifications_user_read_seq", columnList = "user_id, read_at, seq"),
        Index(name = "idx_notifications_user_seq", columnList = "user_id, seq"),
    ],
)
class Notification(
    @Id val id: String,
    @Column(name = "user_id", nullable = false) @JsonIgnore val userId: Long,
    @Column(nullable = false) val type: String,
    @Column(nullable = false) val title: String,
    @Column(nullable = false, columnDefinition = "TEXT") val body: String,
    @Column(nullable = false) val href: String,
    @Column(name = "read_at") var readAt: Instant?,
    @Column(name = "created_at", nullable = false) val createdAt: Instant,
    // 작성 시점 표시 스냅샷. 프론트는 createdAt 으로 상대시간을 만들고 이 값은 fallback.
    @Column(nullable = false) val time: String,
    @JsonIgnore val seq: Long,
)

interface NotificationRepository : JpaRepository<Notification, String> {
    fun findByUserId(userId: Long, pageable: Pageable): Page<Notification>
    fun findByUserIdAndReadAtIsNull(userId: Long, pageable: Pageable): Page<Notification>
    fun countByUserIdAndReadAtIsNull(userId: Long): Long
    fun findByIdAndUserId(id: String, userId: Long): Notification?

    @Modifying
    @Query("update Notification n set n.readAt = :readAt where n.userId = :userId and n.readAt is null")
    fun markAllRead(@Param("userId") userId: Long, @Param("readAt") readAt: Instant): Int

    @Modifying
    @Query("delete from Notification n where n.userId = :userId and n.readAt is not null")
    fun deleteReadByUserId(@Param("userId") userId: Long): Int
}

data class NotificationResponse(
    val id: String,
    val type: String,
    val title: String,
    val body: String,
    val href: String,
    val read: Boolean,
    val readAt: Instant?,
    val createdAt: Instant?,
    val time: String,
)

data class NotificationsResponse(
    val content: List<NotificationResponse>,
    val page: Int,
    val size: Int,
    val totalElements: Long,
    val totalPages: Int,
    val unreadCount: Long,
)

data class NotificationUnreadCountResponse(val unreadCount: Long)

data class NotificationReadAllResponse(val updatedCount: Long, val unreadCount: Long)

data class NotificationDeleteResponse(val deleted: Boolean, val unreadCount: Long)

data class NotificationDeleteReadResponse(val deletedCount: Long, val unreadCount: Long)

fun Notification.toResponse() = NotificationResponse(
    id = id,
    type = type,
    title = title,
    body = body,
    href = href,
    read = readAt != null,
    readAt = readAt,
    createdAt = createdAt,
    time = time,
)

/**
 * 알림 도메인 서비스. 두 가지 책임을 가진다.
 * 1) 도메인 이벤트(댓글/참여)에서 호출되는 알림 생성 helper. 호출자 트랜잭션에 참여한다(notify/notifyUser).
 * 2) 알림 조회/읽음/삭제 비즈니스 정책. Controller 에서 옮겨온 검증·트랜잭션·소유권 확인을 담당한다.
 */
@Service
class NotificationService(private val repo: NotificationRepository) {

    @Transactional(readOnly = true)
    fun getNotifications(userId: Long, page: Int, size: Int, unreadOnly: Boolean): NotificationsResponse {
        validatePageable(page, size)
        val pageable = PageRequest.of(page, size, Sort.by(Sort.Order.desc("seq"), Sort.Order.asc("id")))
        val result = if (unreadOnly) {
            repo.findByUserIdAndReadAtIsNull(userId, pageable)
        } else {
            repo.findByUserId(userId, pageable)
        }
        return NotificationsResponse(
            content = result.content.map { it.toResponse() },
            page = result.number,
            size = result.size,
            totalElements = result.totalElements,
            totalPages = result.totalPages,
            unreadCount = repo.countByUserIdAndReadAtIsNull(userId),
        )
    }

    @Transactional(readOnly = true)
    fun getUnreadCount(userId: Long): NotificationUnreadCountResponse =
        NotificationUnreadCountResponse(repo.countByUserIdAndReadAtIsNull(userId))

    @Transactional
    fun markAsRead(userId: Long, notificationId: String): NotificationResponse {
        val notification = ownedOrNotFound(userId, notificationId)
        if (notification.readAt == null) notification.readAt = Instant.now()
        return notification.toResponse()
    }

    @Transactional
    fun markAllAsRead(userId: Long): NotificationReadAllResponse {
        val updated = repo.markAllRead(userId, Instant.now())
        return NotificationReadAllResponse(updatedCount = updated.toLong(), unreadCount = 0)
    }

    @Transactional
    fun deleteNotification(userId: Long, notificationId: String): NotificationDeleteResponse {
        repo.delete(ownedOrNotFound(userId, notificationId))
        return NotificationDeleteResponse(deleted = true, unreadCount = repo.countByUserIdAndReadAtIsNull(userId))
    }

    @Transactional
    fun deleteReadNotifications(userId: Long): NotificationDeleteReadResponse {
        val deleted = repo.deleteReadByUserId(userId)
        return NotificationDeleteReadResponse(
            deletedCount = deleted.toLong(),
            unreadCount = repo.countByUserIdAndReadAtIsNull(userId),
        )
    }

    private fun ownedOrNotFound(userId: Long, notificationId: String): Notification =
        repo.findByIdAndUserId(notificationId, userId)
            ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "notification $notificationId not found")

    private fun validatePageable(page: Int, size: Int) {
        if (page < 0) throw ResponseStatusException(HttpStatus.BAD_REQUEST, "page must not be negative")
        if (size !in 1..MAX_PAGE_SIZE) {
            throw ResponseStatusException(HttpStatus.BAD_REQUEST, "size must be between 1 and $MAX_PAGE_SIZE")
        }
    }

    fun notify(
        recipientUserId: Long?,
        actorUserId: Long,
        type: String,
        title: String,
        body: String,
        href: String,
    ) {
        if (recipientUserId == null || recipientUserId == actorUserId) return
        notifyUser(recipientUserId, type, title, body, href)
    }

    /**
     * 수신자에게 직접 알림 생성(actor==receiver 여부와 무관). 강제 퇴장처럼 본인에게도 알려야 하는 경우에 쓴다.
     * 같은 트랜잭션에 참여하며 DB 제약 위반은 삼키지 않는다.
     */
    fun notifyUser(
        recipientUserId: Long,
        type: String,
        title: String,
        body: String,
        href: String,
    ) {
        repo.save(
            Notification(
                id = "noti-${UUID.randomUUID()}",
                userId = recipientUserId,
                type = type,
                title = title,
                body = body.trim().take(MAX_BODY),
                href = href,
                readAt = null,
                createdAt = Instant.now(),
                time = "방금 전",
                seq = System.nanoTime(),
            ),
        )
    }

    private companion object {
        const val MAX_BODY = 200
        const val MAX_PAGE_SIZE = 100
    }
}

/** HTTP adapter. 인증 사용자 추출 후 Service 위임과 status code 반환만 담당한다. */
@RestController
@RequestMapping("/api/notifications")
class NotificationController(private val service: NotificationService) {

    @GetMapping
    fun list(
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "20") size: Int,
        @RequestParam(defaultValue = "false") unreadOnly: Boolean,
        @AuthenticationPrincipal user: AuthUser,
    ): NotificationsResponse = service.getNotifications(user.id, page, size, unreadOnly)

    @GetMapping("/unread-count")
    fun unreadCount(@AuthenticationPrincipal user: AuthUser): NotificationUnreadCountResponse =
        service.getUnreadCount(user.id)

    @PostMapping("/{id}/read")
    fun read(@PathVariable id: String, @AuthenticationPrincipal user: AuthUser): NotificationResponse =
        service.markAsRead(user.id, id)

    @PostMapping("/read-all")
    fun readAll(@AuthenticationPrincipal user: AuthUser): NotificationReadAllResponse =
        service.markAllAsRead(user.id)

    @DeleteMapping("/read")
    fun deleteRead(@AuthenticationPrincipal user: AuthUser): NotificationDeleteReadResponse =
        service.deleteReadNotifications(user.id)

    @DeleteMapping("/{id}")
    fun delete(@PathVariable id: String, @AuthenticationPrincipal user: AuthUser): NotificationDeleteResponse =
        service.deleteNotification(user.id, id)
}

File: apps/web/src/data/notifications.ts
// 알림은 백엔드(/api/notifications, 인증 필수)가 source of truth. 사용자별 데이터.
import { apiDelete, apiGet, apiPost } from "@/lib/api";

export type NotificationItem = {
  id: string;
  type: string;
  title: string;
  body: string;
  href: string;
  read: boolean;
  readAt: string | null;
  createdAt?: string | null;
  time: string;
};

export type NotificationsResponse = {
  content: NotificationItem[];
  page: number;
  size: number;
  totalElements: number;
  totalPages: number;
  unreadCount: number;
};

export type NotificationUnreadCountResponse = {
  unreadCount: number;
};

export type NotificationReadAllResponse = {
  updatedCount: number;
  unreadCount: number;
};

export type NotificationDeleteResponse = {
  deleted: boolean;
  unreadCount: number;
};

export type NotificationDeleteReadResponse = {
  deletedCount: number;
  unreadCount: number;
};

// 읽음 처리 후 헤더 badge 가 다시 조회하도록 알리는 이벤트(auth 이벤트와 별개).
// WS 로 unreadCount 가 실려 오면 detail 에 담아 재조회 없이 배지를 갱신한다.
export const NOTIF_EVENT = "dasida-notif";

export type NotifChangedDetail = {
  unreadCount?: number;
};

export function emitNotificationsChanged(detail?: NotifChangedDetail) {
  if (typeof window === "undefined") return;
  window.dispatchEvent(new CustomEvent<NotifChangedDetail>(NOTIF_EVENT, { detail }));
}

export function fetchNotifications(
  page: number,
  size: number,
  unreadOnly: boolean,
  types?: string[],
): Promise<NotificationsResponse> {
  const params = new URLSearchParams({
    page: String(page),
    size: String(size),
    unreadOnly: String(unreadOnly),
  });
  if (types && types.length > 0) params.set("types", types.join(","));
  return apiGet<NotificationsResponse>(`/api/notifications?${params.toString()}`);
}

export function fetchNotificationUnreadCount(): Promise<NotificationUnreadCountResponse> {
  return apiGet<NotificationUnreadCountResponse>("/api/notifications/unread-count");
}

export async function markNotificationRead(id: string): Promise<NotificationItem> {
  const res = await apiPost<NotificationItem>(`/api/notifications/${id}/read`, {});
  emitNotificationsChanged();
  return res;
}

export async function markAllNotificationsRead(): Promise<NotificationReadAllResponse> {
  const res = await apiPost<NotificationReadAllResponse>("/api/notifications/read-all", {});
  emitNotificationsChanged();
  return res;
}

export async function deleteNotification(
  id: string,
): Promise<NotificationDeleteResponse> {
  const res = await apiDelete<NotificationDeleteResponse>(`/api/notifications/${id}`);
  emitNotificationsChanged();
  return res;
}

export async function deleteReadNotifications(
): Promise<NotificationDeleteReadResponse> {
  const res = await apiDelete<NotificationDeleteReadResponse>("/api/notifications/read");
  emitNotificationsChanged();
  return res;
}

/** createdAt 으로 상대시간 라벨 생성. 없으면 저장된 time 스냅샷 fallback. */
export function relativeTime(item: Pick<NotificationItem, "createdAt" | "time">): string {
  if (!item.createdAt) return item.time;
  const then = new Date(item.createdAt).getTime();
  if (Number.isNaN(then)) return item.time;
  const diff = Date.now() - then;
  const min = Math.floor(diff / 60000);
  if (min < 1) return "방금 전";
  if (min < 60) return `${min}분 전`;
  const hours = Math.floor(min / 60);
  if (hours < 24) return `${hours}시간 전`;
  const days = Math.floor(hours / 24);
  if (days < 7) return `${days}일 전`;
  return new Date(then).toLocaleDateString("ko-KR");
}

export function notificationTypeLabel(type: string): string {
  if (type === "CAMPAIGN_JOINED") return "캠페인 참여";
  if (type === "USER_FOLLOWED") return "새 팔로워";
  if (type === "MESSAGE_RECEIVED") return "새 메시지";
  if (type === "POST_COMMENT_CREATED") return "게시글 댓글";
  if (type === "CAMPAIGN_COMMENT_CREATED") return "캠페인 댓글";
  if (type === "POST_LIKED") return "게시글 좋아요";
  if (type === "CAMPAIGN_STATUS_CHANGED") return "캠페인 상태";
  if (type === "REPORT_RESOLVED") return "신고 처리";
  if (type === "CONTENT_HIDDEN") return "콘텐츠 숨김";
  if (type === "CONTENT_RESTORED") return "숨김 해제";
  return "알림";
}

export function isNotificationNavigable(href: string): boolean {
  const trimmed = href.trim();
  return trimmed.startsWith("/") && trimmed.length > 1;
}
